File: subjects/english/power-toeic-ios/Sources/PowerTOEIC/Core/WorkoutBuilder.swift
import Foundation

public enum WorkoutBuilderError: Error, Equatable, Sendable {
    case invalidTotalCount
    case invalidSeed
    case duplicateSkillAllocation(String)
    case missingSkillID
    case invalidAllocation(String)
    case allocationCountsExceedTotal
    case testMustHideSkill
    case reviewMustUseDueSelection
    case presetRequiresSkill(WorkoutMode)
    case weaknessRequiresSkills
}

public struct QuestionHistoryEntry: Codable, Equatable, Sendable {
    public let questionId: String
    public let answeredAt: String

    public init(questionId: String, answeredAt: String) {
        self.questionId = questionId
        self.answeredAt = answeredAt
    }
}

public enum WorkoutBuilder {
    public static func validate(_ recipe: WorkoutRecipe) throws {
        guard recipe.totalCount > 0 else { throw WorkoutBuilderError.invalidTotalCount }
        guard recipe.seed >= 0 else { throw WorkoutBuilderError.invalidSeed }

        var seen = Set<String>()
        var explicitCount = 0
        for allocation in recipe.skillAllocations {
            guard !allocation.skillId.isEmpty else { throw WorkoutBuilderError.missingSkillID }
            guard seen.insert(allocation.skillId).inserted else {
                throw WorkoutBuilderError.duplicateSkillAllocation(allocation.skillId)
            }
            let hasCount = allocation.count != nil
            let hasWeight = allocation.weight != nil
            guard hasCount != hasWeight else {
                throw WorkoutBuilderError.invalidAllocation(allocation.skillId)
            }
            if let count = allocation.count {
                guard count > 0 else { throw WorkoutBuilderError.invalidAllocation(allocation.skillId) }
                explicitCount += count
            }
            if let weight = allocation.weight {
                guard weight.isFinite, weight > 0 else {
                    throw WorkoutBuilderError.invalidAllocation(allocation.skillId)
                }
            }
        }
        guard explicitCount <= recipe.totalCount else {
            throw WorkoutBuilderError.allocationCountsExceedTotal
        }
        if recipe.mode == .test, recipe.labelPolicy != .hideSkill {
            throw WorkoutBuilderError.testMustHideSkill
        }
        if recipe.mode == .review, recipe.selectionPolicy != .reviewDue {
            throw WorkoutBuilderError.reviewMustUseDueSelection
        }
    }

    public static func selectQuestionIDs(
        repository: any QuestionBankRepository,
        recipe: WorkoutRecipe,
        history: [QuestionHistoryEntry] = [],
        reviewQuestionIDs: [String] = []
    ) throws -> [String] {
        try validate(recipe)
        var selected: [String] = []
        var used = Set<String>()
        let reviewSet = Set(reviewQuestionIDs)

        func addFrom(_ questions: [Question], count requestedCount: Int, salt: Int) {
            var remaining = requestedCount
            for question in rankCandidates(questions, history: history, seed: recipe.seed + salt) {
                if selected.count >= recipe.totalCount || remaining <= 0 { break }
                guard used.insert(question.id).inserted else { continue }
                selected.append(question.id)
                remaining -= 1
            }
        }

        if recipe.selectionPolicy == .reviewDue {
            let eligible = try repository.questions().filter { reviewSet.contains($0.id) }
            addFrom(eligible, count: recipe.totalCount, salt: 9_000)
            return selected
        }

        let allocations = allocationCounts(recipe)
        for (index, allocation) in allocations.enumerated() {
            let questions = try repository.questions(skillId: allocation.skillId)
            addFrom(questions, count: allocation.resolvedCount, salt: index + 1)
        }
        if selected.count < recipe.totalCount {
            addFrom(
                try repository.questions(),
                count: recipe.totalCount - selected.count,
                salt: 5_000
            )
        }
        return selected
    }

    public static func presetRecipe(
        _ mode: WorkoutMode,
        skillId: String? = nil,
        totalCount: Int? = nil,
        seed: Int = 1,
        endless: Bool = false
    ) throws -> WorkoutRecipe {
        let defaults: [WorkoutMode: Int] = [
            .quick: 10,
            .training: 30,
            .power: 100,
            .test: 30,
            .review: 30,
        ]
        guard let count = totalCount ?? defaults[mode] else {
            throw WorkoutBuilderError.invalidTotalCount
        }
        if (mode == .training || mode == .power), skillId == nil {
            throw WorkoutBuilderError.presetRequiresSkill(mode)
        }
        let allocations = skillId.map { [SkillAllocation(skillId: $0, count: count)] } ?? []
        let recipe = WorkoutRecipe(
            mode: mode,
            totalCount: count,
            skillAllocations: allocations,
            selectionPolicy: mode == .review ? .reviewDue : .standard,
            labelPolicy: mode == .test ? .hideSkill : .showSkill,
            seed: seed,
            endless: endless
        )
        try validate(recipe)
        return recipe
    }

    public static func weaknessRecipe(
        rankedWeaknesses: [(skillId: String, score: Double)],
        totalCount: Int = 30,
        maxSkills: Int = 3,
        seed: Int = 1
    ) throws -> WorkoutRecipe {
        let eligible = rankedWeaknesses
            .filter { !$0.skillId.isEmpty }
            .prefix(maxSkills)
        guard !eligible.isEmpty else { throw WorkoutBuilderError.weaknessRequiresSkills }
        let count = eligible.count
        let allocations = eligible.enumerated().map { index, entry in
            SkillAllocation(
                skillId: entry.skillId,
                weight: max(1, entry.score.isFinite && entry.score != 0 ? entry.score : Double(count - index))
            )
        }
        let recipe = WorkoutRecipe(
            mode: .weakness,
            totalCount: totalCount,
            skillAllocations: allocations,
            selectionPolicy: .standard,
            labelPolicy: .showSkill,
            seed: seed,
            endless: false
        )
        try validate(recipe)
        return recipe
    }

    private struct ResolvedAllocation {
        let skillId: String
        let weight: Double?
        var resolvedCount: Int
    }

    private static func allocationCounts(_ recipe: WorkoutRecipe) -> [ResolvedAllocation] {
        guard !recipe.skillAllocations.isEmpty else { return [] }
        var result = recipe.skillAllocations.map {
            ResolvedAllocation(skillId: $0.skillId, weight: $0.weight, resolvedCount: $0.count ?? 0)
        }
        var remaining = recipe.totalCount - result.reduce(0) { $0 + $1.resolvedCount }
        let weightedIndices = result.indices.filter { result[$0].weight != nil }
        guard !weightedIndices.isEmpty else { return result }
        let weightTotal = weightedIndices.reduce(0.0) { $0 + (result[$1].weight ?? 0) }
        let weightedBudget = remaining
        for index in weightedIndices {
            let share = Int(floor(Double(weightedBudget) * (result[index].weight ?? 0) / weightTotal))
            result[index].resolvedCount += share
            remaining -= share
        }
        var cursor = 0
        while remaining > 0 {
            let index = weightedIndices[cursor % weightedIndices.count]
            result[index].resolvedCount += 1
            cursor += 1
            remaining -= 1
        }
        return result
    }

    private struct HistoryValue {
        var lastSeen: Int64
        var seenCount: Int
    }

    private static func historyByQuestion(_ history: [QuestionHistoryEntry]) -> [String: HistoryValue] {
        var result: [String: HistoryValue] = [:]
        for (index, entry) in history.enumerated() {
            guard !entry.questionId.isEmpty else { continue }
            let parsed = iso8601Milliseconds(entry.answeredAt) ?? Int64(index)
            if var previous = result[entry.questionId] {
                previous.seenCount += 1
                if parsed >= previous.lastSeen { previous.lastSeen = parsed }
                result[entry.questionId] = previous
            } else {
                result[entry.questionId] = HistoryValue(lastSeen: parsed, seenCount: 1)
            }
        }
        return result
    }

    private static func rankCandidates(
        _ questions: [Question],
        history: [QuestionHistoryEntry],
        seed: Int
    ) -> [Question] {
        let seen = historyByQuestion(history)
        return questions.sorted { lhs, rhs in
            let left = seen[lhs.id]
            let right = seen[rhs.id]
            if left == nil, right != nil { return true }
            if left != nil, right == nil { return false }
            if let left, let right, left.lastSeen != right.lastSeen {
                return left.lastSeen < right.lastSeen
            }
            let leftHash = hashSeed(seed: seed, text: lhs.id)
            let rightHash = hashSeed(seed: seed, text: rhs.id)
            if leftHash != rightHash { return leftHash < rightHash }
            return lhs.id < rhs.id
        }
    }

    private static func hashSeed(seed: Int, text: String) -> UInt32 {
        var hash = UInt32(2_166_136_261) ^ UInt32(truncatingIfNeeded: seed)
        for codeUnit in text.utf16 {
            hash ^= UInt32(codeUnit)
            hash = hash &* 16_777_619
        }
        return hash
    }

    private static func iso8601Milliseconds(_ value: String) -> Int64? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        guard let date = formatter.date(from: value) else { return nil }
        return Int64(date.timeIntervalSince1970 * 1_000)
    }
}

File: subjects/english/power-toeic/js/core/workout-builder.js
const MODES = new Set(['QUICK', 'TRAINING', 'POWER', 'WEAKNESS', 'CUSTOM', 'TEST', 'REVIEW']);
const LABEL_POLICIES = new Set(['show_skill', 'hide_skill']);
const SELECTION_POLICIES = new Set(['standard', 'review_due']);

function freezeRecipe(recipe) {
  return Object.freeze({
    ...recipe,
    skillAllocations: Object.freeze(recipe.skillAllocations.map((entry) => Object.freeze({ ...entry }))),
  });
}

export function createWorkoutRecipe({
  mode,
  totalCount,
  skillAllocations = [],
  selectionPolicy = 'standard',
  labelPolicy = mode === 'TEST' ? 'hide_skill' : 'show_skill',
  seed = 1,
  endless = false,
} = {}) {
  const normalized = {
    mode: String(mode ?? '').toUpperCase(),
    totalCount: Number(totalCount),
    skillAllocations: skillAllocations.map(({ skillId, count, weight }) => ({
      skillId: String(skillId ?? ''),
      ...(count == null ? {} : { count: Number(count) }),
      ...(weight == null ? {} : { weight: Number(weight) }),
    })),
    selectionPolicy,
    labelPolicy,
    seed: Number(seed),
    endless: Boolean(endless),
  };
  validateWorkoutRecipe(normalized);
  return freezeRecipe(normalized);
}

export function validateWorkoutRecipe(recipe) {
  if (!MODES.has(recipe.mode)) throw new Error(`Unsupported workout mode: ${recipe.mode}`);
  if (!Number.isInteger(recipe.totalCount) || recipe.totalCount <= 0) throw new Error('totalCount must be a positive integer');
  if (!Number.isInteger(recipe.seed) || recipe.seed < 0) throw new Error('seed must be a non-negative integer');
  if (!SELECTION_POLICIES.has(recipe.selectionPolicy)) throw new Error(`Unsupported selectionPolicy: ${recipe.selectionPolicy}`);
  if (!LABEL_POLICIES.has(recipe.labelPolicy)) throw new Error(`Unsupported labelPolicy: ${recipe.labelPolicy}`);
  if (!Array.isArray(recipe.skillAllocations)) throw new Error('skillAllocations must be an array');

  const seen = new Set();
  let explicitCount = 0;
  for (const entry of recipe.skillAllocations) {
    if (!entry.skillId) throw new Error('skill allocation requires skillId');
    if (seen.has(entry.skillId)) throw new Error(`Duplicate skill allocation: ${entry.skillId}`);
    seen.add(entry.skillId);
    const hasCount = entry.count != null;
    const hasWeight = entry.weight != null;
    if (hasCount === hasWeight) throw new Error(`Skill ${entry.skillId} must specify exactly one of count or weight`);
    if (hasCount) {
      if (!Number.isInteger(entry.count) || entry.count <= 0) throw new Error(`Invalid count for ${entry.skillId}`);
      explicitCount += entry.count;
    }
    if (hasWeight && (!Number.isFinite(entry.weight) || entry.weight <= 0)) throw new Error(`Invalid weight for ${entry.skillId}`);
  }
  if (explicitCount > recipe.totalCount) throw new Error('Skill allocation counts exceed totalCount');
  if (recipe.mode === 'TEST' && recipe.labelPolicy !== 'hide_skill') throw new Error('TEST mode must hide skill labels');
  if (recipe.mode === 'REVIEW' && recipe.selectionPolicy !== 'review_due') throw new Error('REVIEW mode must use review_due selection');
  return true;
}

function hashSeed(seed, text) {
  let h = (2166136261 ^ seed) >>> 0;
  for (let i = 0; i < text.length; i += 1) {
    h ^= text.charCodeAt(i);
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h >>> 0;
}

function historyByQuestion(attempts = []) {
  const map = new Map();
  attempts.forEach((attempt, index) => {
    const id = attempt.questionId;
    if (!id) return;
    const when = Date.parse(attempt.answeredAt ?? attempt.timestamp ?? '') || index;
    const previous = map.get(id);
    if (!previous || when >= previous.lastSeen) map.set(id, { lastSeen: when, seenCount: (previous?.seenCount ?? 0) + 1 });
    else previous.seenCount += 1;
  });
  return map;
}

function rankCandidates(questions, attempts, seed) {
  const history = historyByQuestion(attempts);
  return [...questions].sort((a, b) => {
    const ah = history.get(a.id);
    const bh = history.get(b.id);
    if (!ah && bh) return -1;
    if (ah && !bh) return 1;
    if (ah && bh && ah.lastSeen !== bh.lastSeen) return ah.lastSeen - bh.lastSeen;
    return hashSeed(seed, a.id) - hashSeed(seed, b.id) || a.id.localeCompare(b.id);
  });
}

function allocationCounts(recipe) {
  if (!recipe.skillAllocations.length) return [];
  const result = recipe.skillAllocations.map((entry) => ({ ...entry, resolvedCount: entry.count ?? 0 }));
  let remaining = recipe.totalCount - result.reduce((sum, entry) => sum + entry.resolvedCount, 0);
  const weighted = result.filter((entry) => entry.weight != null);
  if (!weighted.length) return result;
  const weightTotal = weighted.reduce((sum, entry) => sum + entry.weight, 0);
  const weightedBudget = remaining;
  weighted.forEach((entry) => {
    const share = Math.floor((weightedBudget * entry.weight) / weightTotal);
    entry.resolvedCount += share;
    remaining -= share;
  });
  let cursor = 0;
  while (remaining > 0 && weighted.length) {
    weighted[cursor % weighted.length].resolvedCount += 1;
    cursor += 1;
    remaining -= 1;
  }
  return result;
}

export function selectQuestionIds({ repository, recipe, attempts = [], reviewQuestionIds = [] }) {
  validateWorkoutRecipe(recipe);
  const selected = [];
  const used = new Set();
  const reviewSet = new Set(reviewQuestionIds);
  const addFrom = (questions, count, salt) => {
    for (const question of rankCandidates(questions, attempts, recipe.seed + salt)) {
      if (selected.length >= recipe.totalCount || count <= 0) break;
      if (used.has(question.id)) continue;
      used.add(question.id);
      selected.push(question.id);
      count -= 1;
    }
  };

  if (recipe.selectionPolicy === 'review_due') {
    addFrom(repository.listQuestions().filter((q) => reviewSet.has(q.id)), recipe.totalCount, 9000);
    return Object.freeze(selected);
  }

  const allocations = allocationCounts(recipe);
  allocations.forEach((entry, index) => addFrom(repository.listQuestions({ skillId: entry.skillId }), entry.resolvedCount, index + 1));
  if (selected.length < recipe.totalCount) addFrom(repository.listQuestions(), recipe.totalCount - selected.length, 5000);
  return Object.freeze(selected);
}

export function createPresetRecipe(mode, { skillId = null, totalCount, seed = 1, endless = false } = {}) {
  const upper = String(mode).toUpperCase();
  const defaults = { QUICK: 10, TRAINING: 30, POWER: 100, TEST: 30, REVIEW: 30 };
  const count = totalCount ?? defaults[upper];
  if (!count) throw new Error(`Preset requires totalCount: ${upper}`);
  if ((upper === 'TRAINING' || upper === 'POWER') && !skillId) throw new Error(`${upper} preset requires skillId`);
  return createWorkoutRecipe({
    mode: upper,
    totalCount: count,
    skillAllocations: skillId ? [{ skillId, count }] : [],
    selectionPolicy: upper === 'REVIEW' ? 'review_due' : 'standard',
    labelPolicy: upper === 'TEST' ? 'hide_skill' : 'show_skill',
    seed,
    endless,
  });
}

export function createWeaknessWorkoutRecipe(rankedWeaknesses, { totalCount = 30, maxSkills = 3, seed = 1 } = {}) {
  const eligible = rankedWeaknesses.filter((entry) => entry?.skillId).slice(0, maxSkills);
  if (!eligible.length) throw new Error('Cannot build weakness workout without demonstrated weak skills');
  const allocations = eligible.map((entry, index) => ({
    skillId: entry.skillId,
    weight: Math.max(1, Number(entry.score) || (eligible.length - index)),
  }));
  return createWorkoutRecipe({ mode: 'WEAKNESS', totalCount, skillAllocations: allocations, seed });
}
